Issue #6973 - Setup Request/Response objects for success with RequestLog

joakime opened the pull request:
This is the initial breakdown of PR #6997 into separate PRs (requested by @gregw)

+ Prevents reading of unread/unparsed Request body parameters by RequestLog implementations (existing parameters parsed during dispatch handling are still accessible)
+ Still allows raw Request.getInputStream() and Request.getReader() usage by RequestLog implementations
+ Restores committed response status code. (to simply satisfy BadMessage condition logging)
+ Does not reset committed response headers. (left for future PR)
+ Adding testcase for post-commit response header issue. (currently left disabled)

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #6973 - Setup Request/Response objects for success with RequestLog

+ Prevents reading of Request body parameters
+ Still allows raw Request.getInputStream() and
  Request.getReader() usage
+ Restores committed response status code.
+ Does not rest committed response headers.
+ Adding testcase for post-commit response header
  issue. (currently disabled)

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #6973 - Setup Request/Response objects for success with RequestLog

+ Updating comments

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

gregw commented on jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -1413,6 +1413,20 @@ public String changeSessionId()
+        // Don't allow pulling more parameters
+        _contentParamsExtracted = true;
+
+        // Reset the status code to what was committed
+        MetaData.Response committedResponse = getResponse().getCommittedMetaData();
+        if (committedResponse != null)
+        {
+            getResponse().setStatus(committedResponse.getStatus());
+            // TODO: Reset the response headers to what they were when committed
+        }
```
Let's leave all the restoring of state to a different PR, as it is a different issue

joakime replied:
I'm only restoring response status code, which needs to happen for BadMessage during HttpParser.
There's still a long list of other restoring needed.
See the test cases.

joakime replied:
The part 2 PR is much more comprehensive, this is the minimum in my view.
Example: We have the entire prevention of Request.getInputStream() and Request.getReader() usage during RequestLog as well in part 2. (which wasn't even in the old PR yet)
This is literally what the first commit in the old PR was, before you asked for the rest (isMutable, restoring response headers, preventing post-commit changes, etc) that belongs in part 2.

gregw replied:
I'm thinking we can achieve PR2 with a change to the `_errorSentAndIncludes` field that will prevent mutations rather than needing to restore them.
I'll have a play with that, if you can reduce this PR down to the absolute minimum to fix the reading parameters problem-  ie: move the onCompleted() call to before the request log and then set the parameter extracted boolean within the onCompleted call

gregw commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java:
```diff
@@ -873,7 +873,10 @@ public void onCompleted()
 
         if (_requestLog != null)
+        {
+            _request.onRequestLog();
```
Could we not just move the call to `_request.onCompleted()` to before this line and then avoid creating another event and callback to the request?

joakime replied:
I believe `onCompleted()` is too late, by that point the request / response has been recycled already.

joakime replied:
We also don't want all of the unnecessary steps (like the restoring) to occur if there's no RequestLog declared / active.

gregw replied:
@joakime I believe that we can move the call to `onCompleted` to be just before the request log steps.  I think it is at least worth a try.   It makes sense to not log the request until it's handling is completed.

joakime replied:
Moving the fixing of Request to Request.onComplete() didn't work for the logback test suite.

Moving the fixing of the Request (contentParamsExtracted and status) to Request.onCompleted() breaks logging for session information (they get auth[nz] details out from sessions), and multipart information (file sizes), etc.

Currently, with the PR like it sits here, the RequestLog.log() occurs during HttpChannel.onComplete(), and before Channel idle timeout reset, and Request.onCompleted().
Turns out this location is the perfect place for RequestLog.log() to make sense.

We need another event before Request.onComplete() to fixate the Request/Response.
Or move the exiting behaviors in Request.onComplete() to a new event like Request.recycle() so we can use Request.onComplete() for RequestLog behaviors.
The two events with a RequestLog.log() in between is still needed.

Stated in a different way.

1. HttpChannel.onComplete()
2. Fixate Request/Response.
3. RequestLog.log().
4. Cleanup Request.

joakime replied:
I have a potential solution.
At least one that logback tests seem happy with.

gregw commented:
@joakime have a look at jetty-9.4.x-6973-immutable-once-committed branch.  I think that is a simple/efficient way to make the response immutable once committed.  That way we never need to restore any status nor fields

joakime commented:
> @joakime have a look at jetty-9.4.x-6973-immutable-once-committed branch. I think that is a simple/efficient way to make the response immutable once committed. That way we never need to restore any status nor fields

We still have Request related (session, auth, attributes), and Connection related (remote/local addr) restore of state as well to work out.

gregw commented:
@joakime I'm not inclined to make big changes in 9/10/11 for logging of session/auth/attributes.   Most specifically, I'm don't believe that we should restore session and/or request attributes to what they were when the request was committed. Session are shared by many requests, so that's not even a thing and there are good reasons that a request attribute may change after commit.

So I think we should fix the hard bug (reading params when we should not) and the buglet of mutable status/fields after commit, but I'd dubious there is more we should do prior to jetty-12.

joakime committed:
Issue #6973 - Setup Request/Response objects for success with RequestLog

+ Remove Request.onRequestLog()
+ Move requestlog calling from HttpChannel to Request.onCompleted

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #6973 - Setup Request/Response objects for success with RequestLog

+ address scenario where HttpChannel is null

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>
